File: android/src/main/java/com/reactlibrary/core/DirectionsServiceModule.java
package com.reactlibrary.core;

import androidx.annotation.NonNull;

import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.WritableMap;
import com.google.gson.Gson;
import com.mapsindoors.core.MPDirectionsService;
import com.mapsindoors.core.MPPoint;
import com.mapsindoors.core.errors.MapsIndoorsException;

import java.util.Date;
import java.util.HashMap;
import java.util.UUID;

public class DirectionsServiceModule extends ReactContextBaseJavaModule {

    private HashMap<String, MPDirectionsService> serviceMap = new HashMap<>();
    private static final String NO_DS = "This DirectionsService is not available";

    private final Gson gson = new Gson();
    public DirectionsServiceModule(@NonNull ReactApplicationContext reactContext) {
        super(reactContext);
    }

    @ReactMethod
    public void create(final Promise promise) {
        MPDirectionsService service = new MPDirectionsService();
        String uuid = UUID.randomUUID().toString();
        serviceMap.put(uuid, service);
        promise.resolve(uuid);
    }

    @ReactMethod
    public void addAvoidWayType(String wayType, final String id, final Promise promise) {
        MPDirectionsService service = serviceMap.get(id);
        if (service != null) {
            service.addAvoidWayType(wayType);
            promise.resolve(null);
        } else {
            promise.reject(NO_DS,new MapsIndoorsException((NO_DS)));
        }
    }

    @ReactMethod
    public void clearWayType(final String id, final Promise promise) {
        MPDirectionsService service = serviceMap.get(id);
        if (service != null) {
            service.clearWayType();
            promise.resolve(null);
        } else {
            promise.reject(NO_DS,new MapsIndoorsException((NO_DS)));
        }
    }

    @ReactMethod
    public void setIsDeparture(boolean isDeparture, final String id, final Promise promise) {
        MPDirectionsService service = serviceMap.get(id);
        if (service != null) {
            service.setIsDeparture(isDeparture);
            promise.resolve(null);
        } else {
            promise.reject(NO_DS,new MapsIndoorsException((NO_DS)));
        }
    }

    @ReactMethod
    public void getRoute(String originString, String destinationString, final String id, final Promise promise) {
        MPDirectionsService service = serviceMap.get(id);
        if (service != null) {
            service.setRouteResultListener((route, error) -> {
                WritableMap map = Arguments.createMap();
                if (route != null) {
                    map.putString("route", gson.toJson(route));
                }
                if (error != null) {
                    map.putString("error", gson.toJson(error));
                }
                promise.resolve(map);
            });
            MPPoint origin = gson.fromJson(originString, MPPoint.class);
            MPPoint destination = gson.fromJson(destinationString, MPPoint.class);

            if (origin != null && destination != null) {
                service.query(origin, destination);
            } else {
                promise.reject(new MapsIndoorsException("Origin:" + originString +
                        ", or Destination:" + destinationString + ", are not parsable"));
            }
        } else {
            promise.reject(NO_DS,new MapsIndoorsException((NO_DS)));
        }

    }

    @ReactMethod
    public void setTravelMode(String travelMode, final String id, final Promise promise) {
        MPDirectionsService service = serviceMap.get(id);
        if (service != null) {
            service.setTravelMode(travelMode);
            promise.resolve(null);
        } else {
            promise.reject(NO_DS,new MapsIndoorsException((NO_DS)));
        }
    }

    @ReactMethod
    public void setTime(int time, final String id, final Promise promise) {
        MPDirectionsService service = serviceMap.get(id);
        if (service != null) {
            Date date = new Date();
            date.setTime(time);
            service.setTime(date);
            promise.resolve(null);
        } else {
            promise.reject(NO_DS,new MapsIndoorsException((NO_DS)));
        }

    }

    @NonNull
    @Override
    public String getName() {
        return "DirectionsService";
    }
}

File: ios/core/DirectionsServiceModule.swift
//
//  DirectionsServiceModule.swift
//  react-native-maps-indoors
//
//  Created by Tim Mikkelsen on 02/05/2023.
//

import MapsIndoors
import MapsIndoorsCore
import MapsIndoorsCodable

@objc(DirectionsService)
public class DirectionsServiceModule: NSObject {
    
    var directionsVariables = [String: DirectionsVariables]()
    
    @objc static func requiresMainQueueSetup() -> Bool { return false }
    
    @objc func create(_ resolve: RCTPromiseResolveBlock, rejecter reject: RCTPromiseRejectBlock) {
        let uuid = UUID().uuidString
        let directionsVars = DirectionsVariables()
        directionsVariables[uuid] = directionsVars
        return resolve(uuid)
    }
    
    @objc func addAvoidWayType(_ wayType: String, id: String, resolver resolve: RCTPromiseResolveBlock, rejecter reject: RCTPromiseRejectBlock) {
        guard let directionsVars = directionsVariables[id] else {
            return doReject(reject, message: "directions service not found. Did you call create")
        }
        
        guard let avoidWayType = MPHighway(typeString: wayType) as? MPHighway else {
            return doReject(reject, message: "WayType not found")
        }
        
        directionsVars.wayTypes.append(avoidWayType)
    }
    
    @objc func clearWayType(_ id: String, resolver resolve: RCTPromiseResolveBlock, rejecter reject: RCTPromiseRejectBlock) {
        guard let directionsVars = directionsVariables[id] else {
            return doReject(reject, message: "directions service not found. Did you call create")
        }
        
        directionsVars.wayTypes.removeAll()
    }
    
    @objc func getRoute(_ originString: String, destinationString: String, id: String, resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
        
        guard let directionsVars = directionsVariables[id] else {
            return doReject(reject, message: "directions service not found. Did you call create")
        }
        
        var origin: MPPoint? = nil
        var destination: MPPoint? = nil
        
        do {
            origin = try JSONDecoder().decode(MPPoint.self, from: Data(originString.utf8))
            destination = try JSONDecoder().decode(MPPoint.self, from: Data(destinationString.utf8))
        } catch {
            return doReject(reject, message: "Unable to parse origin or destination point")
        }
        
        if (origin != nil && destination != nil) {
            let query = MPDirectionsQuery(originPoint: origin!, destinationPoint: destination!)
            if (!directionsVars.wayTypes.isEmpty) {
                query.avoidWayTypes = directionsVars.wayTypes
            }
            
            if (directionsVars.date != nil) {
                if (directionsVars.isDeparture) {
                    query.departure = directionsVars.date
                }else {
                    query.arrival = directionsVars.date
                }
            }
            
            query.travelMode = directionsVars.travelMode
            
            Task {
                let route = try await MPMapsIndoors.shared.directionsService.routingWith(query: query)
                if (route != nil) {
                    let routeData = try? JSONEncoder().encode(MPRouteCodable(withRoute: route!))
                    if (routeData != nil) {
                        let routeRoute = String(data: routeData!, encoding: String.Encoding.utf8)
                        var map = Dictionary<String, String>()
                        map["route"] = routeRoute
                        map["error"] = "null"
                        resolve(map)
                    }else {
                        return doReject(reject, message: "Route could not be parsed")
                    }
                }else {
                    return resolve(nil)
                }
            }
        }
        else {
            return doReject(reject, message: "Origin or Destination was null")

        }
    }
    
    @objc func setTravelMode(_ travelMode: String, id: String, resolver resolve: RCTPromiseResolveBlock, rejecter reject: RCTPromiseRejectBlock) {
        
        guard let directionsVars = directionsVariables[id] else {
            return doReject(reject, message: "directions service not found. Did you call create")
        }
        
        var mpTravelMode: MPTravelMode? = nil
        
        switch(travelMode) {
        case "walking":     mpTravelMode = MPTravelMode.walking;
        case "bicycling":   mpTravelMode = MPTravelMode.bicycling;
        case "driving":     mpTravelMode = MPTravelMode.driving;
        case "transit":     mpTravelMode = MPTravelMode.transit;
        default:
            return doReject(reject, message: "Travel mode not found")
        }
        
        directionsVars.travelMode = mpTravelMode!
        resolve(nil)
    }
    
    @objc func setTime(_ time: NSNumber, id: String, resolver resolve: RCTPromiseResolveBlock, rejecter reject: RCTPromiseRejectBlock) {
        
        guard let directionsVars = directionsVariables[id] else {
            return doReject(reject, message: "directions service not found. Did you call create")
        }
        
        directionsVars.date = Date(timeIntervalSince1970: TimeInterval(time.intValue/1000))
        resolve(nil)
    }
    
    @objc func setIsDeparture(_ isDeparture: Bool, id: String, resolver resolve: RCTPromiseResolveBlock, rejecter reject: RCTPromiseRejectBlock) {
        
        guard let directionsVars = directionsVariables[id] else {
            return doReject(reject, message: "directions service not found. Did you call create")
        }
        
        directionsVars.isDeparture = isDeparture
        resolve(nil)
    }
}

public class DirectionsVariables {
    var wayTypes: [MPHighway] = []
    var isDeparture: Bool = true
    var travelMode: MPTravelMode = MPTravelMode.walking
    var date: Date? = nil
}
